amcl/test/basic_localization.py

```diff
@@ -21,61 +22,101 @@ def setUp(self):
         self.target_x = None
         self.target_y = None
         self.target_a = None
+        self.tfBuffer = None
 
-    def tf_cb(self, msg):
-        for t in msg.transforms:
-            if t.header.frame_id == 'map':
-                self.tf = t.transform
-                (a_curr, a_diff) = self.compute_angle_diff()
-                print('Curr:\t %16.6f %16.6f %16.6f' % (self.tf.translation.x, self.tf.translation.y, a_curr))
-                print('Target:\t %16.6f %16.6f %16.6f' % (self.target_x, self.target_y, self.target_a))
-                print('Diff:\t %16.6f %16.6f %16.6f' % (
-                    abs(self.tf.translation.x - self.target_x), abs(self.tf.translation.y - self.target_y), a_diff))
-
-    def compute_angle_diff(self):
+    def print_pose_diff(self):
+        a_curr = self.compute_angle()
+        a_diff = a_curr - self.target_a
+        print('Curr:\t %16.6f %16.6f %16.6f' % (self.tf.translation.x, self.tf.translation.y, a_curr))
+        print('Target:\t %16.6f %16.6f %16.6f' % (self.target_x, self.target_y, self.target_a))
+        print('Diff:\t %16.6f %16.6f %16.6f' % (
+            abs(self.tf.translation.x - self.target_x), abs(self.tf.translation.y - self.target_y), a_diff))
+
+    def get_pose(self):
+        try:
+            tf_stamped = self.tfBuffer.lookup_transform("map", "base_link", rospy.Time())
+            self.tf = tf_stamped.transform
+            self.print_pose_diff()
+        except (tf2.LookupException, tf2.ExtrapolationException):
+            pass
+
+    @staticmethod
+    def wrap_angle_a(angle):
+        """
+        Wrap angle to [-pi, pi)
+        """
+        angle += pi
+        while angle < 0:
+            angle += 2*pi
+        return fmod(angle, 2*pi) - pi
+
+    @staticmethod
+    def wrap_angle_b(angle):
+        """
+        Wrap angle to [0, 2*pi)
+        """
+        while angle < 0:
+            angle += 2*pi
+        return fmod(angle, 2*pi)
+
+    def wrap_angle(self, angle):
+        """
+        Wraps angles according the target angle
+        If the target angle is in the righthand plane, wrap to [-pi, pi)
+        if the target angle is in the lefthand plane, wrap to [0, 2*pi)
+
+        :param angle: Angle to be wrapped
+        :return: wrapped angle
+        """
+        if abs(self.target_a) <= 0.5*pi or abs(self.target_a) >= 1.5*pi:
+            return self.wrap_angle_a(angle)
+        else:
+            return self.wrap_angle_b(angle)
```

Comment:
What is the purpose of this function? why make this distinction?

Reply:
If you want to compare 0,1 pi and 1,9 pi, those are really close, but hard to compare with a simple tolerance check. In that case 0,1 pi and -0,1 pi is easier to compare.

Reply:
My suggestion would be to compute the difference as 1,8 pi and wrap that. Rather than wrapping the values you are comparing. That way you only need wrap_angle_a and you would only need to apply it once